### PR TITLE
Error utils

### PR DESCRIPTION
**Pull request by JoshuaTrujillo15**

### CODE REVIEW

File: src/utils/Errors.huff [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+| sig     || assert failed panic code                                |
+0x4e487b71  000000000000000000000000000000000000000000000000000000001
+
+*/
+#define macro ASSERT() = takes (1) returns (0) {
+    // takes:               // [condition]
+    do_not_panic            // [do_not_panic_jumpdest, condition]
+    jumpi                   // []
+    [ASSERT_FALSE]          // [assert_false]
+    PANIC()                 // []
```
Comment (abigger87) [inline src/utils/Errors.huff:98]: Nice re-usage 👍

File: src/utils/Errors.huff [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+/* Bubbles up revert data if call failed. Call directly after `call`, `staticcall`, `delegatecall`. */
+#define macro BUBBLE_UP_IF_FAILED() = takes (1) returns (0) {
+    // takes:       // [call_succeeded]
+    call_succeeded  // [call_succeeded_jumpdest, call_succeeded]
+    jumpi           // []
+    returndatasize  // [returndatasize]
+    0x00            // [memory_offset, returndatasize]
+    returndatasize  // [returndatasize, memory_offset, returndatasize]
+    dup2            // [returndata_offset, returndatasize, memory_offset]
```
Comment (abigger87) [inline src/utils/Errors.huff:110]: Missing a `returndatasize` at the end of the stack comment here I believe

**Review by abigger87 — CHANGES_REQUESTED** [review]
Small Nits

File: src/utils/Errors.huff [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#define constant EMPTY_ARRAY_POP = 0x31
+#define constant ARRAY_OUT_OF_BOUNDS = 0x32
+#define constant MEMORY_TOO_LARGE = 0x41
+#define constant UNINITIALIZED_FUNCTION_POINTER = 0x51
+
+/*
+
+Solidity Require. Error `string` MUST be no greater than 32 bytes.
+
+MEMORY LAYOUT WHEN THROWN
```
Comment (clabby) [inline src/utils/Errors.huff:40]: I believe strings are encoded as the pointer to the length, length, and then the left-padded contents.

ex:
```
| sig    | message length offset                                   | message length                                          | message "revert"                                        |
0x08c379a 000000000000000000000000000000000000000000000000000000020 000000000000000000000000000000000000000000000000000000006 726576657274000000000000000000000000000000000000000000000
```

See: https://docs.soliditylang.org/en/v0.8.15/control-structures.html#revert

Comment (devtooligan) [reply src/utils/Errors.huff:40]: lol i think both of you guys are missing some zeroes

Comment (clabby) [reply src/utils/Errors.huff:40]: > lol i think both of you guys are missing some zeroes

Definitely was lmao, fixed. good catch 🤝

File: src/utils/Errors.huff (hunk 1 above)
Comment (JoshuaTrujillo15) [reply src/utils/Errors.huff:101]: ty :metal:

File: src/utils/Errors.huff (hunk 2 above)
Comment (JoshuaTrujillo15) [reply src/utils/Errors.huff]: fixed

File: src/utils/Errors.huff (hunk 3 above)
Comment (JoshuaTrujillo15) [reply src/utils/Errors.huff:40]: fixed, ty!

**Review by abigger87 — APPROVED** [review]
LGTM

**Pull request merged**